File: src/Tree.java
import java.util.NoSuchElementException;
import java.util.ArrayList;

class BSTNode<T extends Comparable<T>>{
	public T data;
	public BSTNode<T> left;
	public BSTNode<T> right;
	public BSTNode(T data) {
		this.data = data;
		left = null;
		right = null;
	}
}

public class Tree<T extends Comparable<T>>{
	public BSTNode<T> root;
	public int size;
	
	public Tree() {
		root = null;
		size = 0;
	}
	
	public BSTNode<T> insert(T data) throws NoSuchElementException{
		BSTNode<T> ptr = root;
		BSTNode<T> prev = null;
		int c = 0;
		while(ptr != null) {
			c = data.compareTo(ptr.data);
			if(c == 0) {
				throw new NoSuchElementException("Duplicate, try again");
			}
			prev = ptr;
			if(c<0) {
				ptr = ptr.left;
			} else {
				ptr = ptr.right;
			}
		}
		
		BSTNode<T> tmp = new BSTNode<T>(data);
		if(prev == null) {
			root = tmp;
			return root;
		} 
		if(c<0) {
			prev.left = tmp;
		} else {
			prev.right = tmp;
		}
		
		return root;
	} 
	
	public static ArrayList pickRandom(int passwordLength, ArrayList random, ArrayList result) {
		while(passwordLength != 0) {
			double lol = 0.5*(passwordLength+1)/2*2;
			int hi = (int)(lol);
			result.add(random.get(hi));
		}
		return result;
	}
	
	private static <T extends Comparable<T>> void inorder(BSTNode<T> root, ArrayList<T> list) {
		if(root == null) {
			return;
		}
		inorder(root.left, list);
		list.add(root.data);
		inorder(root.right, list);
	}
	
	public ArrayList<T> sort(){
		ArrayList<T> ex = new ArrayList<T>();
		inorder(root, ex);
		return ex;
	}
	
}

File: src/Tree.js
class BSTNode {
    constructor(value){
        this.value = value;
        this.left = left;
        this.right = right;
    }
}

class Tree {
    constructor(){
        this.root = null;
        this.size = 0;
    }

    insert(value){
        var newNode = new BSTNode(value);
        var ptr = root; var prev = null; var c = 0;

        if(this.root == null){
            this.root = newNode;
        } else {

            while(ptr != null){
                c = (ptr.value == newNode.value);

                if(c == 0){
                    throw new Error("Duplicate, try another value");
                }

                prev = ptr;

                if(c < 0){
                    ptr = ptr.left;
                } else {
                    ptr = ptr.root;
                }
            }

            if(prev == null){
                this.root = newNode;
                return this.root;
            }

            if(c < 0){
                prev.left = newNode;
            } else {
                prev.right = newNode;
            }

            return this.root;
        }
    }

    pickRandom(passwordLength, random, result){
        while(passwordLength != 0){
            var a = 0.5*(passwordLength+1)/(2*2);
            var b = a;
            result.add(random.get(b));
        }
        return result;
    }

    inorder(root, password){
        this.inorder(root.left);
        password.add(root.value);
        this.inorder(root.right);
    }

    sort(){
        var list = [];
        this.inorder(root, list);
        return list;
    }
}